pkg/restore/cdclog/kv.go

```diff
@@ -0,0 +1,359 @@
+			incrementalBits := typeBitsLength - kvcodec.tbl.Meta().AutoRandomBits
+			hasSignBit := !mysql.HasUnsignedFlag(col.Flag)
+			if hasSignBit {
+				incrementalBits--
+			}
+			_ = kvcodec.tbl.RebaseAutoID(kvcodec.se, value.GetInt64()&((1<<incrementalBits)-1), false, autoid.AutoRandomType)
+		}
+		if isAutoIncCol {
+			_ = kvcodec.tbl.RebaseAutoID(kvcodec.se, value.GetInt64(), false, autoid.AutoIncrementType)
+		}
+	}
+
+	if TableHasAutoRowID(kvcodec.tbl.Meta()) {
+		j := columnPermutation[len(cols)]
+		if j >= 0 && j < len(row) {
+			value, err = table.CastValue(kvcodec.se, row[j], extraHandleColumnInfo, false, false)
+		} else {
+			value, err = types.NewIntDatum(rowID), nil
+		}
+		if err != nil {
+			return nil, err
+		}
+		record = append(record, value)
+		_ = kvcodec.tbl.RebaseAutoID(kvcodec.se, value.GetInt64(), false, autoid.RowIDAllocType)
+	}
+	_, err = kvcodec.tbl.AddRecord(kvcodec.se, record)
+	if err != nil {
+		log.Error("kv add Record failed",
+			zap.Array("originalRow", rowArrayMarshaler(row)),
+			zap.Array("convertedRow", rowArrayMarshaler(record)),
+			zap.Error(err),
+		)
+		return nil, errors.Trace(err)
+	}
+
+	pairs := kvcodec.se.takeKvPairs()
+	kvcodec.recordCache = record[:0]
+	return KvPairs(pairs), nil
+}
+
+// RemoveRecord encode a row of data into KV pairs.
+func (kvcodec *tableKVEncoder) RemoveRecord(
+	row []types.Datum,
+	rowID int64,
+	columnPermutation []int,
+) (Row, error) {
+	cols := kvcodec.tbl.Cols()
+
+	var value types.Datum
+	var err error
+
+	record := kvcodec.recordCache
+	if record == nil {
+		record = make([]types.Datum, 0, len(cols)+1)
+	}
+
+	for i, col := range cols {
+		j := columnPermutation[i]
+		isAutoIncCol := mysql.HasAutoIncrementFlag(col.Flag)
+		switch {
+		case j >= 0 && j < len(row):
+			value, err = table.CastValue(kvcodec.se, row[j], col.ToInfo(), false, false)
+			if err == nil {
+				err = col.HandleBadNull(&value, kvcodec.se.vars.StmtCtx)
+			}
+		case isAutoIncCol:
+			// we still need a conversion, e.g. to catch overflow with a TINYINT column.
+			value, err = table.CastValue(kvcodec.se, types.NewIntDatum(rowID), col.ToInfo(), false, false)
+		default:
+			value, err = table.GetColDefaultValue(kvcodec.se, col.ToInfo())
+		}
+		if err != nil {
+			return nil, err
+		}
+		record = append(record, value)
+	}
+	err = kvcodec.tbl.RemoveRecord(kvcodec.se, kv.IntHandle(rowID), record)
+	if err != nil {
+		log.Error("kv remove record failed",
+			zap.Array("originalRow", rowArrayMarshaler(row)),
+			zap.Array("convertedRow", rowArrayMarshaler(record)),
+			zap.Error(err),
+		)
+		return nil, errors.Trace(err)
+	}
+
+	pairs := kvcodec.se.takeKvPairs()
+	kvcodec.recordCache = record[:0]
+	return KvPairs(pairs), nil
+}
+
+// ClassifyAndAppend split KvPairs to data rows and index rows.
+func (kvs KvPairs) ClassifyAndAppend(
+	data *Rows,
+	dataChecksum *KVChecksum,
+	indices *Rows,
+	indexChecksum *KVChecksum,
+) {
+	dataKVs := (*data).(KvPairs)
```

Comment:
How to implement the `Append` method?🤔️

Reply:
Maybe we could implement `Row` and `Rows` for `*KvPair`...? 🤔

Reply:
I'm not sure about what you mean, could you give the signature of `Append`?

Reply:
The `KvPair` is a typo to `KvPairs`, sorry about it. 😿

Let me explain:

We can make a new method on `Rows`:
```
	// Append append a KV pair to the collection.
	Append(pair KvPair)
```

And we can implement `Rows` and `Row` for `*KvPairs`(instead of `KvPairs`), so we can:
```
func (kvs *KvPairs) Append(pair KvPair) {
	*kvs = append(*kvs, pair) 
}
```

Reply:
The advantages of this are the internal `Rows` implementation won't be leaked to the outer world. However, it also introduces the two-level pointer overhead. If this isn't acceptable or you believe the abstraction leakage doesn't matter, we can also ignore this for now.